### DISCUSSION TITLE
Save Jupyter notebook with interactive scene embedded

### DISCUSSION THREAD

**Discussion by CsatiZoltan** [pyvista/pyvista#4809] [Q&A]
Hi,

For plotting in a Jupyter notebook, I use the recommended backend, `'trame'`:
```python
import pyvista
pyvista.set_jupyter_backend('trame')
```
I can export a scene into an [interactive HTML file](https://github.com/pyvista/pyvista/discussions/1542#discussioncomment-1047907) by clicking on the "Export scene as HTML" button in the scene. However, I would like to export the whole notebook like this. In Jupyter, *File --> Save and Export Notebook As... --> HTML* does not render the PyVista widget, showing only its textual representation:
```
Widget(value="<iframe src=...)
```
Is it possible to export the notebook as an HTML file with the **interactive** scene embedded? If not, I would need to fall back to the `'static'` backend.

**Comment by johannes-mueller** [comment]
My question goes in the same direction. Up to now I could using the `panel` backend embed interactive 3d plots into my static html documentation (see [this example](https://pylife.readthedocs.io/en/stable/demos/hotspot_beam.html) created from a jupyter notebook). Now that the `panel` backend is gone what is the way to do that. I tried with `trame` and `client` but that simply does not output anything.

**Reply by CsatiZoltan** [reply]
Nice example, thanks for the link!

**Reply by adeak** [reply]
@johannes-mueller I'm not familiar with these mechanics, but when the backends were replaced in https://github.com/pyvista/pyvista/pull/4569 Bane created a demo repo with dynamic plots at https://github.com/pyvista/pyvista-docs-dynamic, don't know if this might be helpful. We also have interactive plots on the main page, see `index.rst`, but those use `jupyter-execute` with a static backend, so I have no idea what magic is involved :)

**Reply by CsatiZoltan** [reply]
Thank you, @adeak. Concerning *pyvista-docs-dynamic*, the generation process is passed to *sphinx-gallery*, as appears in the [config file](https://github.com/pyvista/pyvista-docs-dynamic/blob/main/doc/conf.py) (see also *pyvista/plotting/utilities/sphinx_gallery.py*). It would be nice if exporting to an interactive HTML would not require setting up a Sphinx project.

Your other pointer, [index.rst](https://github.com/pyvista/pyvista/blob/main/doc/source/index.rst#brief-examples) uses the `jupyter-execute` directive from [*jupyter-sphinx*](https://jupyter-sphinx.readthedocs.io/en/latest/), so it also seems to require Sphinx.

I am not against using Sphinx if that is the only way to solve the problem. However, in the first case, an annotated Python file was turned to an interactive page, while in the second example, code snippets were embedded into an RST file. I would like to publish from a Jupyter notebook (.ipynb file), just as @johannes-mueller [did it](https://github.com/boschresearch/pylife/blob/8cc8a8247d1553e3645872bce452b98967c80bfd/demos/hotspot_beam.ipynb). Do you have any ideas?

**Reply by johannes-mueller** [reply]
@adeak thanks for the hint to `index.rst`. That shows that it works in principle. However, I was not able to replicate the interactive plots in my environment. I copypasted the `.. jupyter-execute` blocks from `index.rst` to an `.rst` file in my Sphinx doctree and I end up with a static image.

My jupyter notebooks show the interactive plots nicely *when run in a jupyter server*. However, when I render static html from the notbooks using nbsphinx  the best I can seem to get are static pngs. Shall I file an issue that the method how to plot interactive plots with nbsphinx or jupyter-sphinx should be documented more clearly?

**Reply by adeak** [reply]
It took me a full local doc build, but finally I've realised that there's no magic: on `main` the main page is no longer interactive, as you'd expect from the static backend :facepalm: Cf. [stable (currently 0.41) docs](https://docs.pyvista.org/version/0.41/index.html) with [dev docs](https://docs.pyvista.org/version/dev/index.html). Sorry for the red herring @johannes-mueller :(

I can only confirm now that `Plotter.export_html()` (or `Plotter.export_vtksz()` + `trame_vtk.tools`' `static_viewer.html` offline viewer as used by [our offline viewer directive](https://github.com/pyvista/pyvista/blob/66fc5f86a579791b0c6ead2909e2696edef4598d/pyvista/ext/viewer_directive.py#L57-L59)) work. I don't know how to get these to work with notebooks.

> Shall I file an issue that the method how to plot interactive plots with nbsphinx or jupyter-sphinx should be documented more clearly?

That would definitely be helpful (unless there's already an open issue/discussion about this), thanks. With the old backends recently deprecated/removed it would be nice to have easy to find guidelines about using the new system. And to come back to the question of @CsatiZoltan, doing this outside sphinx (but inside jupyter) would also be valuable. (But as I said I don't know how the old backends worked in the first place.)

**Reply by adeak** [reply]
One more thing, although this is probably not it:

> I tried with `trame` and `client` but that simply does not output anything.

Depending on what "does not output anything" means, I've seen people report that they see empty (white) figures when opening the exported .html files. It might be the case that one has to have a proper HTTP server running on localhost serving the html file (can be as simple as `python -m http.server`). When I locally tested `Plotter.export_vtksz()` for my previous comment, I saw the same thing. Firefox complained about a CORS violation when I opened the HTML file directly, but serving it on localhost through `http.server` fixed the render.

**Comment by eigenvivek** [comment]
I'm also interested in this feature. I use [Quarto](https://quarto.org/docs/get-started/) (Jupyter -> HTML) to build documentation for my package, and would love if PyVista plots would automatically export as HTML and render in the browser.

Interestingly, plots made in Plotly already seem to do this.

**Comment by CsatiZoltan** [comment]
In the meantime, I found [K3D](https://github.com/K3D-tools/K3D-jupyter), which can also embed interactive content into HTML. Here is an [example for VTK](https://k3d-jupyter.org/gallery/showcase/vtk-cutter.html), which might be relevant for PyVista. I hope you can make something with this information, I myself have no experience with Javascript and OpenGL.

**Comment by hakonanes** [comment]
Hi all, I'm yet another user who hosts online docs with interactive plots using PyVista. I specifically use the panel backend in notebooks that are included in a Sphinx documentation via nbsphinx. See the bottom plot in [this notebook](https://kikuchipy.org/en/stable/tutorials/visualizing_patterns.html#Plot-master-patterns) for an example. I have the same experience as @johannes-mueller.

This works out-of-the-box with the versions of PyVista that support panel (screenshot from online docs):

![bilde](https://github.com/pyvista/pyvista/assets/12139781/f95fc70c-99fe-4039-8b6a-903ac6f498c3)

Replacing panel with trame does not work (screenshot from locally built Sphinx docs):

![bilde](https://github.com/pyvista/pyvista/assets/12139781/d574c615-2309-49d3-9403-c45690ef1600)

The Norwegian text reads something like: "Cannot connect. Firefox could not connect to the server on localhost:34533... Try again."

**Reply by banesullivan** [reply]
@hakonanes, I think what you want to do requires what @ChristosT is implementing in https://github.com/pyvista/pyvista/pull/4938

**Reply by hakonanes** [reply]
Thank you, @banesullivan (and @ChristosT). It does indeed.

We use both Sphinx-Gallery (shorter examples) and nbsphinx (longer tutorials) in many projects. Scripts for a Sphinx-Gallery aren't as suited for longer tutorials as a notebook is, I feel. So it would be fantastic if PyVista could support interactive plots using both with Sphinx!

**Comment by banesullivan** [comment]
This should be feasible using much of the same interface we've developed for:

- #4569
- https://kitware.github.io/vtk-js/examples/OfflineLocalView.html
- https://pyvista.github.io/pyvista-docs-dynamic/
- https://github.com/pyvista/pyvista/pull/4938

cc @ChristosT and @jourdain might have some ideas

At the end of the day, we just need PyVista's `Widget` class (below) to implement a this exporting in whatever mechanism Jupyter expects for saving/serializing the state. As far as I can tell, this should not be a heavy lift.

https://github.com/pyvista/pyvista/blob/5e1b0ec2b8681c6b04b857ee30e670fdf46b97e6/pyvista/trame/jupyter.py#L70

**Reply by ChristosT** [reply]
There is a similar issue here 

https://github.com/ansys/pyansys-geometry/pull/826

The main idea is to export the scene and create an iframe manually 
```python
scene = pl.export_html(filename=None)

from IPython.display import display, HTML
html_str = scene.getvalue().replace('"','&quot;')
iframe=f"""
<iframe
    srcdoc="{html_str}"
    width='100%'
    height='500'
</iframe>
"""
HTML(iframe)
```
In the referenced issue I tried to just hide the ouput/input cells to hide this snippet but using a widget is indeed a much better solution !

@banesullivan Do you think we can have another `mode` (named `html` maybe?) in https://github.com/pyvista/pyvista/blob/5e1b0ec2b8681c6b04b857ee30e670fdf46b97e6/pyvista/trame/jupyter.py#L212 

where we can return a `Widget` with its `iframe` defined similarly to the snippet above ?

**Reply by banesullivan** [reply]
> Do you think we can have another mode (named html maybe?) in

Definitely! I think `html` would be an appropriate name for this. 

@ChristosT, do you want to try to include this in #4938? Or perhaps tackle this after that lands?

**Reply by banesullivan** [reply]
(I'm happy to assist here, I just lack the bandwidth to be the primary person on this)

**Reply by ChristosT** [reply]
Thanks ! I will look into this, but maybe in a separate pull request since #4938 takes time fine-tune correctly. Besides, I think that it could be done independently from that work since there is no use of the `DynamicScrapper` or the `plot-directive` here . We just want a standalone exported scene that can run without the trame server.

I will ping you when I have something.

**Reply by banesullivan** [reply]
Ah you're totally right that this could be its own, smaller, targeted PR regardless of the dynamic scraper work. Definitely ping me if you get something... I may also take a stab at it if I get motivated on of these evenings

**Reply by banesullivan** [reply]
Work in progress here: https://github.com/pyvista/pyvista/pull/5168

**Reply by johannes-mueller** [reply]
As #5168 has been merged now is even in the current release, I tried it out and it seems to work ([see here for an example](https://pylife.readthedocs.io/en/2.1/demos/hotspot_beam.html)).

However, there is one caveat if you build this from jupyter notebooks in your code base that when not used to build the html would not use the `html` backend.

The solution/workaround I have found is a little bit hacky as it monkey patches preprocessing code into `nbconvert.Exporter`.  Anyway, for those who look for a solution [here](https://github.com/boschresearch/pylife/blob/7a7305a3aceb3f3c6f3a167af3809be178d8cde6/docs/conf.py#L52-L72) something you could use for inspiration.  There's probably room for improvement.

**Reply by ChristosT** [reply]
You could also set it as environmental variable see here https://github.com/ansys/pyansys-geometry/pull/826/commits/aac6175f736e720db0b8f8d68202ccc3df9ce23b 

Unfortunately, I am not sure where is the appropriate place to place it for `nbconvert` .
